Add `pathlib` brain to handle `.parents` inference

deepyaman opened the pull request:
<!--

Thank you for submitting a PR to astroid!

To ease our work reviewing your PR, do make sure to mark the complete the following boxes.

-->

## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description

The `PurePath.parents` sequence supports slices since Python 3.10. This brain allows pylint to understand that passing an index (as opposed to a slice) to `PurePath.parents` will still yield a path, not a tuple.

Given `test_index.py`:

```python
from pathlib import Path


def _main() -> None:
    current_path = Path().resolve()
    parent_path = current_path.parents[2].resolve()
    print(f"{current_path=}")
    print(f"{parent_path=}")


if __name__ == "__main__":
    _main()
```

Running pylint yields:

```
(astroid-3.10) WASH-178551-C02X31K9JHD4:customerone deepyaman_datta$ pylint test_index.py --disable missing-module-docstring

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

Given `test_slice.py`:

```python
from pathlib import Path


def _main() -> None:
    current_path = Path().resolve()
    parent_paths = [p.resolve() for p in current_path.parents[:2]]
    print(f"{current_path=}")
    print(f"{parent_paths=}")


if __name__ == "__main__":
    _main()
```

Running pylint yields:

```
(astroid-3.10) WASH-178551-C02X31K9JHD4:customerone deepyaman_datta$ pylint test_slice.py --disable missing-module-docstring

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|     | :sparkles: New feature |
|     | :hammer: Refactoring   |
|     | :scroll: Docs          |

## Related Issue

<!--
If this PR fixes a particular issue, use the following to automatically close that issue
once this PR gets merged:

Closes #XXX
-->
Closes with a test case https://github.com/PyCQA/pylint/issues/5783
Partially addresses https://github.com/PyCQA/pylint/issues/5832

DanielNoord commented on astroid/brain/brain_pathlib.py:
```diff
@@ -0,0 +1,36 @@
+from astroid.const import PY310_PLUS
+from astroid.exceptions import UseInferenceDefault
+from astroid.manager import AstroidManager
+from astroid.nodes.node_classes import Attribute, Slice, Subscript
```
Could you import `from astroid import nodes` here? The `nodes` module shouldn't depend on this brain so we should be fine 😄

DanielNoord commented on astroid/brain/brain_pathlib.py:
```diff
@@ -0,0 +1,36 @@
+from astroid.nodes.node_ng import NodeNG
+
+
+def _looks_like_parents_subscript(node: NodeNG) -> bool:
```
It has been quite some time that I looked into this myself, but I believe `register_transform` makes it so that only the class passed as first argument is pass to this function.

Perhaps you could do this and run pytest:
```python
if not isinstance(node, nodes.Subscript):
    breakpoint()
```

If there is no break then we can update this to be `nodes.Subscript` instead.

deepyaman replied:
Got it, tested it and (of course) it's as you all said!

DanielNoord commented on tests/unittest_brain_pathlib.py:
```diff
@@ -0,0 +1,42 @@
+from astroid.const import PY310_PLUS
+
+if not PY310_PLUS:
+    pytest.skip(
```
Can you add tests for <3.10 as well? So just mark the tests for 3.10+ as such and add some tests for the old behaviour.

deepyaman replied:
Done!

DanielNoord commented on tests/unittest_brain_pathlib.py:
```diff
@@ -0,0 +1,42 @@
+    )
+
+
+def test_inference_parents_subscript_index():
```
```suggestion
def test_inference_parents_subscript_index() -> None:
```

DanielNoord commented on tests/unittest_brain_pathlib.py:
```diff
@@ -0,0 +1,42 @@
+    assert inferred.qname() == "pathlib.Path"
+
+
+def test_inference_parents_subscript_slice():
```
```suggestion
def test_inference_parents_subscript_slice() -> None:
```

DanielNoord commented on tests/unittest_brain_pathlib.py:
```diff
@@ -0,0 +1,42 @@
+    parent_path
+    """
+    )
+    inferred = next(name_node.infer())
```
Could you use `list` here and assert that `len(list) == 1`. We want to make sure that there is no ambiguity in the inference here and `next()` does not tell us that.

deepyaman replied:
Done! Used `.inferred()` instead of `.infer()` and checked the length of that, because that's what astroid/tests/unittest_brain_dataclasses.py does, although I'm not entirely sure what the difference is (`.inferred()` leverages an inference cache?).

DanielNoord commented on tests/unittest_brain_pathlib.py:
```diff
@@ -0,0 +1,42 @@
+    parent_path
+    """
+    )
+    inferred = next(name_node.infer())
```
Could you use `list` here and assert that `len(list) == 1`. We want to make sure that there is no ambiguity in the inference here and `next()` does not tell us that.

DanielNoord commented on tests/unittest_brain_pathlib.py:
```diff
@@ -0,0 +1,42 @@
+import pytest
```
Just for "totality" could you also add a test for the inference of `parents` itself?

deepyaman replied:
I feel like I'm almost there, but it's not currently working for:

```py
from pathlib import Path


def _main() -> None:
    current_path = Path().resolve()
    path_parents = current_path.parents
    parent_path = path_parents[2].resolve()
    print(f"{current_path=}")
    print(f"{parent_path=}")


if __name__ == "__main__":
    _main()
```

It fails with the same `test.py:7:18: E1101: Instance of 'tuple' has no 'resolve' member (no-member)`.

Seems to be why my attempt at adding a nicer-looking test failed:

```py
def test_inference_parents_subscript_index() -> None:
    """Test inference of ``pathlib.Path.parents``, accessed by index."""
    parents, path = astroid.extract_node(
        """
    from pathlib import Path

    current_path = Path().resolve()
    path_parents = current_path.parents
    path_parents  #@
    path_parents[2]  #@
    """
    )
    inferred = parents.inferred()
    assert len(inferred) == 1
    assert isinstance(inferred[0], bases.Instance)
    assert inferred[0].qname() == "pathlib._PathParents"

    inferred = path.inferred()
    assert len(inferred) == 1
    if PY310_PLUS:
        assert isinstance(inferred[0], bases.Instance)
        assert inferred[0].qname() == "pathlib.Path"
    else:
        assert inferred[0] is Uninferable
```

DanielNoord commented on astroid/brain/brain_pathlib.py:
```diff
@@ -0,0 +1,36 @@
+
+def infer_parents_subscript(
+    subscript_node: Subscript, ctx: Optional[context.InferenceContext] = None
+) -> Iterator[bases.Instance]:
```
Perhaps we should add a docstring to explain why this is needed, or a comment between L33 and L34.

deepyaman replied:
Comment added!

DanielNoord reviewed (commented):
Thanks @deepyaman, left some initial comments!

I also updated your OP not to directly close the `pylint` issue. We tend to fix issues in `astroid` and then add a regression test to `pylint` to fully close the issue. If you want you can make a PR with a regression test for `pylint` as well!

deepyaman commented:
@DanielNoord Thank you for the detailed feedback! I will try and implement the requested changes/feedback over the next few days and get back to you. :)

cdce8p commented on astroid/brain/brain_pathlib.py:
```diff
@@ -0,0 +1,36 @@
+
+
+def _looks_like_parents_subscript(node: NodeNG) -> bool:
+    return (
+        isinstance(node, Subscript)
+        and isinstance(node.value, Attribute)
+        and node.value.attrname == "parents"
+    )
```
This isn't enough, it will also match unintended code. Just an example. I would suggest adding something like it as a test.
```py
name_node = astroid.extract_node(
    """\
class A:
    parents = 42

c = A()
error = c.parents[2]
error
"""
)
inferred = next(name_node.infer())
print(inferred)
```

deepyaman replied:
Done!

cdce8p commented on astroid/brain/brain_pathlib.py:
```diff
@@ -0,0 +1,36 @@
+    if isinstance(subscript_node.slice, Slice):
+        raise UseInferenceDefault
+
+    yield from subscript_node.value.expr.infer(context=ctx)
```
As mentioned earlier, it's not quite enough to check that `attrname == "parents"`.
A rough idea what might work instead
```py
    inferred = next(subscript_node.value.expr.infer(context=ctx))
    if not (
        isinstance(inferred, bases.Instance)
        and isinstance(inferred._proxied, nodes.ClassDef)
        and inferred.qname().startswith("pathlib.")
        and inferred.qname().endswith("Path")
    ):
        raise UseInferenceDefault
    if isinstance(subscript_node.slice, Slice):
        raise UseInferenceDefault

    return iter([inferred])
```

DanielNoord replied:
I think the check is somewhat unnecessary because `subscript_node.value` is `nodes.Attribute` due to `_looks_like_parents_subscript`.
However, we don't really "know" this when we are in this function and therefore once we turn on `mypy` it will (rightly) complain that `qname` is not defined on `value`. I think it makes sense to add these `isinstance` checks as well.

deepyaman replied:
@cdce8p @DanielNoord I have updated this to add the `isinstance` checks to my implementation.

cdce8p commented on astroid/brain/brain_pathlib.py:
```diff
@@ -0,0 +1,36 @@
+    yield from subscript_node.value.expr.infer(context=ctx)
+
+
+if PY310_PLUS:
+    AstroidManager().register_transform(
+        Subscript, inference_tip(infer_parents_subscript), _looks_like_parents_subscript
+    )
```
```suggestion
AstroidManager().register_transform(
    Subscript, inference_tip(infer_parents_subscript), _looks_like_parents_subscript
)
```
The subscript isn't limited to `3.10`. Just `[2:]`.

deepyaman replied:
As you mentioned above, the result pre-3.10 is to always `Uninferable`, which is also OK; added a comment to explain.

deepyaman commented on tests/unittest_brain_pathlib.py:
```diff
@@ -0,0 +1,63 @@
+    assert inferred.qname() == "builtins.tuple"
+
+
+def test_inference_parents_subscript_not_path():
```
```suggestion
def test_inference_parents_subscript_not_path() -> None:
```

deepyaman commented on tests/unittest_brain_pathlib.py:
```diff
@@ -0,0 +1,42 @@
+    assert inferred.qname() == "pathlib.Path"
+
+
+def test_inference_parents_subscript_slice():
```
Added a note in the implementation, as well as adding a branch for the test case if Python <3.10.

deepyaman commented on tests/unittest_brain_pathlib.py:
```diff
@@ -0,0 +1,84 @@
+        assert isinstance(inferred[0], bases.Instance)
+        assert inferred[0].qname() == "pathlib.Path"
+    else:
+        assert inferred[0] is Uninferable
```
`tuple` should not be correct; when passed an integral index (as opposed to a slice), you get a single path.

DanielNoord replied:
No, but since this PR changes the behaviour here shouldn't we try and infer the correct value? Or do you think that is outside of the scope of this PR?

deepyaman replied:
@DanielNoord Sorry for the delay in taking a look at this. I've removed the `if PY310_PLUS` condition, and the brain tip seems to be testing fine for older Python versions.

deepyaman commented on astroid/brain/brain_pathlib.py:
```diff
@@ -0,0 +1,42 @@
+    yield from subscript_node.value.expr.infer(context=ctx)
+
+
+if PY310_PLUS:
+    # Before adding support for slices to the parents sequence in Python
+    # 3.10, the return value for ``pathlib._PathParents.__getitem__`` is
+    # always ``Uninferable``, so we only apply the brain tip since 3.10.
+    AstroidManager().register_transform(
+        nodes.Subscript,
+        inference_tip(infer_parents_subscript),
+        _looks_like_parents_subscript,
+    )
```
```suggestion
AstroidManager().register_transform(
    nodes.Subscript,
    inference_tip(infer_parents_subscript),
    _looks_like_parents_subscript,
)
```

deepyaman commented on tests/unittest_brain_pathlib.py:
```diff
@@ -0,0 +1,84 @@
+    )
+    inferred = name_node.inferred()
+    assert len(inferred) == 1
+    if PY310_PLUS:
+        assert isinstance(inferred[0], bases.Instance)
+        assert inferred[0].qname() == "pathlib.Path"
+    else:
+        assert inferred[0] is Uninferable
```
```suggestion
    assert isinstance(inferred[0], bases.Instance)
    assert inferred[0].qname() == "pathlib.Path"
```

deepyaman commented on ChangeLog:
```diff
@@ -22,6 +22,10 @@ Release date: TBA
 
   Closes #1410
 
+* Add ``pathlib`` brain to handle ``pathlib.PurePath.parents`` inference in Python 3.10+.
```
```suggestion
* Add ``pathlib`` brain to handle ``pathlib.PurePath.parents`` inference.
```

deepyaman commented:
@DanielNoord @cdce8p Sorry for the delay in addressing your comments, but I believe I've done it now.

I'm still not sure how to resolve this scenario:

```python
(astroid-3.10) WASH-178551-C02X31K9JHD4:astroid deepyaman_datta$ cat test.py 
import astroid

code = """from pathlib import Path
current_path = Path().resolve()
path_parents = current_path.parents
path_parents[2]
"""


module = astroid.parse(code)
node = module.body[-1].value
print(node.inferred())
(astroid-3.10) WASH-178551-C02X31K9JHD4:astroid deepyaman_datta$ python test.py 
[<Instance of builtins.tuple at 0x140247194163456>]
```

Even if I can make the inference tip activate on a `Name` node, I can't access the contextual information I need to determine the type. Should I try and make it return `Uninferable` instead?

I'm also not completely sure why I'm getting the Flake8 issues even after updating to latest `main`.

DanielNoord commented:
> @DanielNoord @cdce8p Sorry for the delay in addressing your comments, but I believe I've done it now.
> 
> I'm still not sure how to resolve this scenario:
> 
> ```python
> (astroid-3.10) WASH-178551-C02X31K9JHD4:astroid deepyaman_datta$ cat test.py 
> import astroid
> 
> code = """from pathlib import Path
> current_path = Path().resolve()
> path_parents = current_path.parents
> path_parents[2]
> """
> 
> 
> module = astroid.parse(code)
> node = module.body[-1].value
> print(node.inferred())
> (astroid-3.10) WASH-178551-C02X31K9JHD4:astroid deepyaman_datta$ python test.py 
> [<Instance of builtins.tuple at 0x140247194163456>]
> ```
> 
> Even if I can make the inference tip activate on a `Name` node, I can't access the contextual information I need to determine the type. Should I try and make it return `Uninferable` instead?

What contextual information are you referring to? I think `uninferable` is probably better than an incorrect inference.

> I'm also not completely sure why I'm getting the Flake8 issues even after updating to latest `main`.

Seems like our CI is broken for flake8. We'll need to change some stuff to fix those issues.

Pierre-Sassoulas commented:
> Seems like our CI is broken for flake8. We'll need to change some stuff to fix those issues.

I think it's due to ``additional_dependencies: [flake8-bugbear,`` where bugbear version is not set. Interestingly it's the same check than what we just added in pylint 2.14.

deepyaman commented:
> What contextual information are you referring to? I think `uninferable` is probably better than an incorrect inference.

@DanielNoord If I change `_looks_like_parents_subscript` to also trigger on `isinstance(node.value, nodes.Name)`:

```python
def _looks_like_parents_subscript(node: nodes.Subscript) -> bool:
    return isinstance(node.value, nodes.Attribute) and node.value.attrname == "parents" or isinstance(node.value, nodes.Name)
```

Everything looks good inside `infer_parents_subscript`:

```python
(Pdb) subscript_node
<Subscript l.4 at 0x7f8ceb29ffd0>
(Pdb) subscript_node.value
<Name.path_parents l.4 at 0x7f8ceb29fa90>
(Pdb) value = next(subscript_node.value.infer())
(Pdb) value
<Instance of pathlib._PathParents at 0x140243218564960>
(Pdb) isinstance(value, bases.Instance) and isinstance(value._proxied, nodes.ClassDef) and value.qname() == "pathlib._PathParents"
True
(Pdb) subscript_node.slice
<Const.int l.4 at 0x7f8ceb29fbe0>
```

However, I can't figure out any way to infer the type of `current_path` from this point, which is what I'd normally use to set the return type of the `.parents` subscript. There's nothing I can find like `subscript_node.value.expr`.

However, I'm not sure if it's right to handle `Name` nodes explicitly like this, or to make them `Uninferable`, since it seems (in my head, I don't understand `astroid` in depth to actually know) like there would be so many instances where passing a `Name` node/assigning to an intermediate variable would lose a lot of contextual type information. I couldn't find other code handling `Name` nodes so explicitly within `astroid`, at a glance.

deepyaman commented:
> > What contextual information are you referring to? I think `uninferable` is probably better than an incorrect inference.
> 
> @DanielNoord If I change `_looks_like_parents_subscript` to also trigger on `isinstance(node.value, nodes.Name)`:
> 
> ```python
> def _looks_like_parents_subscript(node: nodes.Subscript) -> bool:
>     return isinstance(node.value, nodes.Attribute) and node.value.attrname == "parents" or isinstance(node.value, nodes.Name)
> ```
> 
> Everything looks good inside `infer_parents_subscript`:
> 
> ```python
> (Pdb) subscript_node
> <Subscript l.4 at 0x7f8ceb29ffd0>
> (Pdb) subscript_node.value
> <Name.path_parents l.4 at 0x7f8ceb29fa90>
> (Pdb) value = next(subscript_node.value.infer())
> (Pdb) value
> <Instance of pathlib._PathParents at 0x140243218564960>
> (Pdb) isinstance(value, bases.Instance) and isinstance(value._proxied, nodes.ClassDef) and value.qname() == "pathlib._PathParents"
> True
> (Pdb) subscript_node.slice
> <Const.int l.4 at 0x7f8ceb29fbe0>
> ```
> 
> However, I can't figure out any way to infer the type of `current_path` from this point, which is what I'd normally use to set the return type of the `.parents` subscript. There's nothing I can find like `subscript_node.value.expr`.
> 
> However, I'm not sure if it's right to handle `Name` nodes explicitly like this, or to make them `Uninferable`, since it seems (in my head, I don't understand `astroid` in depth to actually know) like there would be so many instances where passing a `Name` node/assigning to an intermediate variable would lose a lot of contextual type information. I couldn't find other code handling `Name` nodes so explicitly within `astroid`, at a glance.

Hi @DanielNoord @cdce8p! I was wondering if you may have some guidance on how to proceed with this? Would love to close out this issue; just not sure what's the best approach to resolving the above issue.

DanielNoord commented:
@deepyaman Sorry for letting this rest so long. I have been quite busy and haven't had the time to look into this enough. I hope to find some time for it this weekend. Again, sorry for the long delay!

DanielNoord commented:
@deepyaman Sorry for letting this rest so long. I have been quite busy and haven't had the time to look into this enough. I hope to find some time for it this weekend. Again, sorry for the long delay!

deepyaman commented:
> @deepyaman Sorry for letting this rest so long. I have been quite busy and haven't had the time to look into this enough. I hope to find some time for it this weekend. Again, sorry for the long delay!

No problem, and thanks again for taking the time to guide me through this!

DanielNoord reviewed (approved):
@deepyaman Sorry this took so long to review. When I finally got to it I was in so deep that it seemed easier to just push the fixes I found.

Sorry for taking over your PR with prior notice like that.

In this new iteration we now also correctly infer an index lookup. I do this by creating an inference of the normal `ClassDef` for `Path` and then instantiating an instance from it.
I am not completely sure `Uninferable` for a `Slice` is correct on <3.9. I'll let another maintainer comment on that: do we infer `Uninferable` for code that would raise warnings?

I'll re-ask @cdce8p for a review since he had some valuable input in the first round.

deepyaman commented on tests/unittest_brain_pathlib.py:
```diff
@@ -72,7 +77,7 @@ class A:
 
     c = A()
-    error = c.parents[2]
+    error = c.parents[:2]
```
Did this have to be changed?

DanielNoord replied:
Originally I hadn't fixed the normal index look up, so this was not really testing anything:
The brain returned `Uninferable` and the non-brain would return `Uninferable`. So we weren't actually testing if this doesn't actually enter the brain. By using the slice we can see that this doesn't trigger the brain as it would give a different result for `PY3.10+`.

Now that I fixed the brain to work with index as well the change is perhaps not necessary anymore.

deepyaman commented on astroid/brain/brain_pathlib.py:
```diff
@@ -2,36 +2,50 @@
-        raise UseInferenceDefault
+        return False
+
+    return True
```
Maybe just:

```suggestion
    return (
        isinstance(value, bases.Instance)
        and isinstance(value._proxied, nodes.ClassDef)
        and value.qname() == "pathlib._PathParents"
    )
```

DanielNoord replied:
yeah that's probably better! Committed!

deepyaman reviewed (commented):
@DanielNoord No problem! Thanks for resolving it!

cdce8p commented on astroid/brain/brain_pathlib.py:
```diff
@@ -0,0 +1,51 @@
+from astroid.exceptions import InferenceError, UseInferenceDefault
+from astroid.manager import AstroidManager
+
+PATH = extract_node(
+    """from pathlib import Path
+Path
+"""
+)
+PATH_CLASSDEF: nodes.ClassDef = next(PATH.infer())
```
Inference calls are non-trivial. I'm don't think it's a good idea to do them at import. Better to only define the code template as global const and call `extract_node` (or even `_extract_single_node`) in `infer_parents_subscript`.

DanielNoord replied:
But it isn't it better to use the actual `ClassDef`? You're more experienced here, but this seems to pass tests and doesn't this give us more information than a constructed mimic of `Path`?

cdce8p replied:
It's probably fine to use `extract_node` and `infer` to get the `ClassDef`, just don't do it at Module level / import time.

As example, this is how the `typing` brain does something similar
https://github.com/PyCQA/astroid/blob/e57ac42bca80c105bbdad4bae61637b39288a47b/astroid/brain/brain_typing.py#L99-L103
https://github.com/PyCQA/astroid/blob/e57ac42bca80c105bbdad4bae61637b39288a47b/astroid/brain/brain_typing.py#L167-L168

DanielNoord replied:
I'll refactor!

cdce8p commented on astroid/brain/brain_pathlib.py:
```diff
@@ -0,0 +1,51 @@
+
+
+def _looks_like_parents_subscript(node: nodes.Subscript) -> bool:
+    if not isinstance(node.value, nodes.Name) and not (
+        isinstance(node.value, nodes.Attribute) and node.value.attrname == "parents"
+    ):
```
```suggestion
    if not (
        isinstance(node.value, nodes.Name)
        or isinstance(node.value, nodes.Attribute)
        and node.value.attrname == "parents"
    ):
```
`not ... and not (... and ...)` is quite complicated and difficult to follow.

cdce8p commented on astroid/brain/brain_pathlib.py:
```diff
@@ -0,0 +1,51 @@
+    subscript_node: nodes.Subscript, ctx: context.InferenceContext | None = None
+) -> Iterator[bases.Instance]:
+    if isinstance(subscript_node.slice, nodes.Const):
+        return iter((PATH_CLASSDEF.instantiate_class(),))
```
`_extract_single_node` should be done here.

cdce8p commented on ChangeLog:
```diff
@@ -39,6 +43,7 @@ Release date: TBA
 * The Qt brain now correctly treats calling ``.disconnect()`` (with no
   arguments) on a slot as valid.
 
+
```
```suggestion
```

cdce8p commented:
@DanielNoord I moved `extract_node` inside `infer_parents_subscript` as well. We should try to do as little computation as possible in the global scope. With `iter`, I switched the one-item tuple with a list. Although correct, it's quite easy to accidentally remove the trailing comma or overlook it. A list is a bit safer there.

cdce8p reviewed (approved):
Good to merge once the CI checks pass.

DanielNoord commented:
> @DanielNoord I moved `extract_node` inside `infer_parents_subscript` as well. We should try to do as little computation as possible in the global scope. With `iter`, I switched the one-item tuple with a list. Although correct, it's quite easy to accidentally remove the trailing comma or overlook it. A list is a bit safer there.

Thanks! I'm still finding my way through the `astroid` brains. I'll keep the global computation in mind 😄

cdce8p commented:
> I'll keep the global computation in mind 😄

Off-topic: There is an open proposal for lazy imports in Python: [PEP 690](https://peps.python.org/pep-0690/). Thankfully opt-in, but nevertheless will be quite challenging as it is to make pylint and astroid compatible.

DanielNoord commented:
Thanks @deepyaman for the initial PR and thanks to @cdce8p for the reviews!

Happy to finally see this merged!
